### PR TITLE
Core: Adding block styles via editor.BlockEdit filter

### PR DESCRIPTION
**Pull request by retrofox**
## Description

This PR inserts the `<BlockStyles />` component to the Inspector Controls hooking a function to the `editor.BlockEdit` instead of hardcoding. It doesn't add anything new to the API just an improvement in terms of functionality.

It allows to handler properly the components inner the Inspector Controls via the hooks, since right now the BlockStyles locates always at the top and it isn't possible to deal with it, because, as I said before, it's hardcoded 😢 .

## How has this been tested?
* Apply the changes and confirm that `<BlockStyles />` shows as usual for these blocks that have defined style variations.
* Also, you can add a new component at the top of the Inserter Controls defining the priority f the bound function to the filter. It needs to be more than 10:

```es6
import { addFilter } from '@wordpress/hooks';
import { Fragment } from '@wordpress/element';
import { InspectorControls } from '@wordpress/block-editor';

const myEditComponent = OriginalBlockEdit => props => {
	return (
		<Fragment>
			<InspectorControls>
				<div>This is at the top!</div>
			</InspectorControls>
			<OriginalBlockEdit { ...props } />
		</Fragment>
	);
};

addFilter( 'editor.BlockEdit', 'my-block/editComponent', myEditComponent, 20 );
```

Screen Shot 2020-07-16 at 4 18 32 PM

## Screenshots 

Screen Shot 2020-07-16 at 3 00 15 PM

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Review by marekhrabe — COMMENTED** [review]
Described usage works as advertised! Styles are not hardcoded and now go in with a hook. 

I have one tiny concern, outside of my scope of knowledge. Essentially, in a `block-inspector/index.js`, the `BlockStyles` are removed. This means they no longer show up at all. 

To bring them back, a new file was created `hooks/block-styles.js`. It hooks into the right filter and adds them under the same conditions as before. This file is imported in `hooks/index.js` so it always loads together with other similar code. All good, in a browser.

What I have noticed is `hooks/index.js` has a also a native variant — `index.native.js`. This is where my knowledge falls short and there are two variants of what could happen:

1. we have removed hardcoded styles from the `BlockInspector` and never brought them back, because the `hooks/block-styles.js` is not imported from `hooks/index.native.js`
2. styles work differently on mobile and they don't use `BlockInspector`. 

I have no idea which one is true, as I have no experience with the native side of the editor. However, considering that `hooks/index.native.js` imports almost everything that the standard `hooks/index.js` does, I think we might need to add the import too.

Any idea who can we ping to review this? It's probably super easy to confirm for anybody having the experience with mobile.

**Comment by geriux** [comment]
Hey there 👋 

> styles work differently on mobile and they don't use BlockInspector.

That is correct, we don't use the `BlockInspector` on mobile.

> However, considering that hooks/index.native.js imports almost everything that the standard hooks/index.js does, I think we might need to add the import too.

For this case is not needed to import it in the `.native` side.

Thank you for double checking that it wouldn't break anything for mobile!

**Comment by marekhrabe** [comment]
That's perfect, @geriux! Thanks for your confirmation! And with that…

**Review by marekhrabe — APPROVED** [review]

File: packages/block-editor/src/hooks/block-styles.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+			const blockType =
+				selectedBlockClientId && getBlockType( selectedBlockName );
+			const blockStyles =
+				selectedBlockClientId && getBlockStyles( selectedBlockName );
+			return {
+				blockType,
+				selectedBlockClientId,
+				hasBlockStyles: blockStyles && blockStyles.length > 0,
+			};
+		} ),
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/block-styles.js]: An opportunity to refactor to use useSelect?

**Comment by youknowriad** [comment]
No objections from here.

File: packages/block-editor/src/hooks/block-styles.js (hunk 1 above)
Comment (retrofox) [reply packages/block-editor/src/hooks/block-styles.js]: Yes, for sure. Why do you prefer using hooks instead of composition?

Comment (youknowriad) [reply packages/block-editor/src/hooks/block-styles.js]: It avoids an unnecessary block wrapper and I believe it results in a better API as you know exactly what is an external API of the component (prop) and what is just internal.

Comment (retrofox) [reply packages/block-editor/src/hooks/block-styles.js]: Awesome. Thanks a lot. Addressed here 467de0e1a2d329b64a136353253c29484f4a7d75

**Comment by getdave** [comment]
@retrofox will you be looking to merge this?

**Comment by retrofox** [comment]
> @retrofox will you be looking to merge this?

yeah, let me rebase it and then will try to ship it. Thanks for the reminder, Dave 🙇‍♂️

**Comment by retrofox** [comment]
Rebased and ready for a new review? cc @getdave

**Comment by getdave** [comment]
@retrofox I'll add this to my list to review on Friday.

**Comment by retrofox** [comment]
> No objections from here.

HI @youknowriad 👋 
I wonder if you can take a look and ✅ if you consider good enough? I've just tested and it works as expected.
Thanks in advance.

File: packages/block-editor/src/hooks/block-styles.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		);
+	},
+	'withInspectorControls'
+);
+
+addFilter(
+	'editor.BlockEdit',
+	'core/editor/block-styles/with-inspector-controls',
+	withInspectorControls
+);
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/block-styles.js:63]: Do we need a priority or something to tell it to load first? how is this ensured?

Anyway, this approach looks good to me.

Comment (retrofox) [reply packages/block-editor/src/hooks/block-styles.js:63]: yes, that's a good idea. let me increase to `20` to give it a priority in its position.

**Comment by Mamaduka** [comment]
I was about to create PR for this and then came across this one. Thanks, @retrofox.

Is there a reason for using `selectedBlockClientId` instead of `clientId` and `blockName` from props?

P.S. Let me know if I can help to get this PR merged.

**Comment by retrofox** [comment]
> I was about to create PR for this and then came across this one. Thanks, @retrofox.

🙇‍♂️  

> Is there a reason for using `selectedBlockClientId` instead of `clientId` and `blockName` from props?

No, actually let me update the PR. It needs a rebase, too.

> P.S. Let me know if I can help to get this PR merged.

Sounds like a nice plan. 🤝

**Comment by retrofox** [comment]
> P.S. Let me know if I can help to get this PR merged.

Ready for a new 👁️ round, @Mamaduka 🙇‍♂️

File: packages/block-editor/src/hooks/block-styles.js [hunk 3]
```diff
@@ -18,29 +18,18 @@ import DefaultStylePicker from '../components/default-style-picker';
...
-				return {
-					blockType:
-						blockClientId && getBlockType( selectedBlockName ),
-					selectedBlockClientId: blockClientId,
-					hasBlockStyles: blockStyles && blockStyles.length > 0,
-				};
-			},
-			[]
+		const hasBlockStyles = useSelect(
+			( select ) => select( blocksStore ).getBlockStyles( name ),
```
Comment (Mamaduka) [inline packages/block-editor/src/hooks/block-styles.js]: Might be better if we calculate the boolean derived value here. One might expect `hasBlockStyles` to be boolean, based on the naming convention.

Something like this:
```js
( select ) => select( blocksStore ).getBlockStyles( name ).length > 0
```

P.S. Sorry for nitpicking.

**Comment by Mamaduka** [comment]
Hi, @retrofox

Tested this, and it works as expected.

File: packages/block-editor/src/hooks/block-styles.js (hunk 3 above)
Comment (retrofox) [reply packages/block-editor/src/hooks/block-styles.js]: Don't say sorry at all :-). Sounds great to me. Going to update the line below, too.

Comment (retrofox) [reply packages/block-editor/src/hooks/block-styles.js]: handled here https://github.com/WordPress/gutenberg/pull/23993/commits/acf9861a85be5be270521136a96dac0ce11e7dac

**Comment by retrofox** [comment]
> I believe this should be `! hasBlockStyles`.
> 
> Everything else looks good!

🤦 fixed!. Thanks :-)

**Review by getdave — COMMENTED** [review]
Nice simple change which seems to work well 👍 

I tested this by temporarily adding the following to the `core/image` block's `edit.js` file

```js
const myEditComponent = ( OriginalBlockEdit ) => ( props ) => {
	return (
		<Fragment>
			<InspectorControls>
				<div>This is at the top!</div>
			</InspectorControls>
			<OriginalBlockEdit { ...props } />
		</Fragment>
	);
};

addFilter( 'editor.BlockEdit', 'core/image', myEditComponent, 20 );
```

I saw the text `This is at the top!` appear above the block styles in the block sidebar.

I should be possible to write simple e2e tests for this by using the pattern shown in `packages/e2e-tests/specs/editor/plugins/align-hook.test.js` which activates a Plugin for testing. Inside the Plugin, you can do the `addFilter` and add some random text above/below the block styles using the hook. Then you can use e2e tests to assert on the presence of the text.

Might be over the top though.

Otherwise this seems to test well for me.

**Pull request closed without merging**